pkg/queryfrontend/cache_splitter.go

```diff
@@ -8,17 +8,32 @@ import (
 	"time"
 
 	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+
+	"github.com/thanos-io/thanos/pkg/compact/downsample"
 )
 
-// constSplitter is a utility for using a constant split interval when determining cache keys.
-type constSplitter time.Duration
+// thanosCacheSplitter is a utility for using split interval when determining cache keys.
+type thanosCacheSplitter struct {
+	interval    time.Duration
+	resolutions []int64
+}
+
+func newThanosCacheSplitter(interval time.Duration) thanosCacheSplitter {
+	return thanosCacheSplitter{
+		interval:    interval,
+		resolutions: []int64{downsample.ResLevel2, downsample.ResLevel1, downsample.ResLevel0},
+	}
+}
 
 // TODO(yeya24): Add other request params as request key.
 // GenerateCacheKey generates a cache key based on the Request and interval.
-func (t constSplitter) GenerateCacheKey(_ string, r queryrange.Request) string {
-	currentInterval := r.GetStart() / time.Duration(t).Milliseconds()
+func (t thanosCacheSplitter) GenerateCacheKey(_ string, r queryrange.Request) string {
```

Comment:
It's not splitter anymore, rather `cacheKeyGenerator` ? (:

Reply:
Yes, that's better